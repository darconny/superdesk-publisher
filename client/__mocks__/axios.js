// import axios from "axios"

// // Mock out all top level functions, such as get, put, delete and post:
// jest.mock("axios")

// axios.get.mockImplementation(url => {
//     if (url === 'www.example.com') {
//         return Promise.resolve({ data: {...} })
//     } else {

//     }
// })

"use strict";
module.exports = {
  get: url => {
    let data = {};

    if (url.includes("tenants")) {
      data = {
        _embedded: {
          _items: [
            {
              id: 1,
              subdomain: "tenant1",
              domain_name: "sourcefabric.org",
              name: "TENANT1",
              code: "tenant1",
              organization: {
                id: 1,
                name: "test",
                code: "0gjdk8"
              },
              created_at: "2017-06-30T09:36:00+00:00",
              updated_at: "2019-04-16T19:45:19+00:00",
              enabled: true,
              theme_name: "swp/daily-theme",
              amp_enabled: true,
              output_channel: null,
              articles_count: 140,
              fbia_enabled: false,
              paywall_enabled: false
            },
            {
              id: 2,
              subdomain: "tenant2",
              domain_name: "sourcefabric.org",
              name: "TENANT2",
              code: "tenant2",
              organization: {
                id: 1,
                name: "test",
                code: "0gjdk8"
              },
              created_at: "2017-06-30T09:36:00+00:00",
              updated_at: "2019-04-16T19:45:19+00:00",
              enabled: true,
              theme_name: "swp/daily-theme",
              amp_enabled: true,
              output_channel: null,
              articles_count: 140,
              fbia_enabled: false,
              paywall_enabled: false
            }
          ]
        }
      };
    }

    if (url.includes("content/routes")) {
      data = {
        _embedded: {
          _items: [
            {
              id: 1,
              name: "route1"
            },
            {
              id: 2,
              name: "route2"
            }
          ]
        }
      };
    }

    if (url.includes("content/lists")) {
      data = {
        _embedded: {
          _items: [
            {
              id: 1,
              name: "content list 1",
              type: "manual"
            },
            {
              id: 2,
              name: "content list 2",
              type: "manual"
            }
          ]
        }
      };
    }

    if (url.includes("packages/seo")) {
      data = {
        meta_title: null,
        meta_description: null,
        og_title: null,
        og_description: null,
        twitter_title: null,
        twitter_description: null
      };
    }

    // console.log("\x1b[44m\x1b[37m%s\x1b[0m", "API URL: " + url);
    return Promise.resolve({ data: data });
  },

  post: url => {
    let data = {};

    if (url.includes("organization/articles/related")) {
      data = {
        related_article_items: [
          { tenants: [], title: "test related 1" },
          {
            tenants: [
              {
                code: "eif0ca"
              }
            ],
            title: "test related 2"
          }
        ]
      };
    }

    if (url.includes("auth/superdesk")) {
      data = {
        token: { api_key: "asdfasdf" }
      };
    }

    if (url.includes("preview/package/generate_token")) {
      data = {
        preview_url: "https://sourcefabric.org/preview"
      };
    }

    if (url.includes("organization/destinations")) {
      data = {};
    }

    // console.log("\x1b[44m\x1b[37m%s\x1b[0m", "API URL: " + url);
    return Promise.resolve({ data: data });
  }
};
