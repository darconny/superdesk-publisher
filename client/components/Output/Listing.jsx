import React, { forwardRef } from "react";
import PropTypes from "prop-types";

import helpers from "../../services/helpers.js";
import VirtualizedList from "../generic/VirtualizedList";
import ArticleItem from "./ArticleItem";
import Store from "./Store";

class Listing extends React.Component {
  static contextType = Store;

  constructor(props) {
    super(props);
    this._isMounted = false;
    this.queryLimit = 20;

    this.state = {
      articles: {
        items: [],
        page: 0,
        totalPages: 1,
        loading: true,
        itemSize: 56
      }
    };
  }

  componentDidMount() {
    this._isMounted = true;
    this.context.publisher.setToken().then(this._queryArticles);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  buildQueryParams = () => {
    let queryParams = {
      page: this.state.articles.page + 1,
      limit: this.queryLimit,
      "status[]": []
    };

    // let route = scope.buildRouteParams();
    // let tenant = scope.buildTenantParams();
    // let universal = scope.buildUniversalParams();

    // queryParams = Object.assign(queryParams, universal);

    if (this.props.type === "incoming") {
      queryParams["status[]"] = ["new"];
      queryParams["sorting[created_at]"] = "desc";
    } else {
      queryParams["status[]"] = ["published", "unpublished"];
      // queryParams["tenant[]"] = tenant.length ? tenant : undefined;
      // queryParams["route[]"] = route.length ? route : undefined;
      // queryParams.published_before = scope.filters.publishedBefore;
      // queryParams.published_after = scope.filters.publishedAfter;
      queryParams["sorting[updated_at]"] = "desc";
    }
    return queryParams;
  };

  _queryArticles = () => {
    let queryParams = this.buildQueryParams();
    let articles = { ...this.state.articles };
    articles.loading = true;
    this.setState({ articles });

    this.context.publisher
      .queryMonitoringArticles(queryParams)
      .then(response => {
        let newArticles = response._embedded._items;

        if (this.context.selectedList === "published") {
          newArticles.forEach(item => {
            item.comments_count = helpers.countComments(item.articles);
            item.page_views_count = helpers.countPageViews(item.articles);
            item.articles.forEach(item => {
              if (item.route && item.status == "published" && item.tenant) {
                let tenantUrl = item.tenant.subdomain
                  ? item.tenant.subdomain + "." + item.tenant.domain_name
                  : item.tenant.domain_name;
                item.live_url = "http://" + tenantUrl + item._links.online.href;
              }
            });
          });
        }

        let articlesCounts = { ...this.context.articlesCounts };
        articlesCounts[this.props.type] = response.total;
        this.context.actions.setArticlesCounts(articlesCounts);

        articles.items = [...articles.items, ...newArticles];
        articles.loading = false;
        articles.page = response.page;
        articles.totalPages = response.pages;

        this.setState({ articles });
      })
      .catch(err => {
        articles.loading = false;
        this.setState({ articles });
        this.context.notify.error("Cannot load articles");
      });
  };

  render() {
    if (this.props.hide) return null;

    const innerElementType = forwardRef(({ style, ...rest }, ref) => (
      <div
        ref={ref}
        className="sd-list-item-group sd-shadow--z2"
        style={{
          ...style,
          width: "calc(100% - 4.8rem)"
        }}
        {...rest}
      />
    ));

    return (
      <div className="sd-column-box__main-column">
        {this.state.articles.items.length ? (
          <VirtualizedList
            hasNextPage={
              this.state.articles.totalPages > this.state.articles.page
                ? true
                : false
            }
            isNextPageLoading={this.state.articles.loading}
            loadNextPage={this._queryArticles}
            items={this.state.articles.items}
            itemSize={this.state.articles.itemSize}
            ItemRenderer={ArticleItem}
            innerElementType={innerElementType}
          />
        ) : (
          <div className="panel-info">
            <div className="panel-info__icon">
              <i className="big-icon--text"></i>
            </div>
            <h3 className="panel-info__heading">
              No {this.props.type} articles.
            </h3>
          </div>
        )}
      </div>
    );
  }
}

Listing.propTypes = {
  type: PropTypes.string.isRequired,
  hide: PropTypes.bool
};

export default Listing;
