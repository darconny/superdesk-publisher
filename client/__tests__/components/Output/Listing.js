import React from "react";
import Listing from "../../../components/Output/Listing";
import { render, waitForElement } from "@testing-library/react";
import axios from "axios";
import Store from "../../../components/Output/Store";

import Publisher from "../../../__mocks__/publisher";

const publisher = new Publisher();

jest.mock("react-virtualized/styles.css", () => {
  return {};
});
jest.mock("axios");
jest.mock("react-select", () => props => "div");

let api = () => { };

api.users = {};
api.users.query = function () {
  return new Promise((resolve, reject) => {
    resolve({
      _items: [{ is_author: true, display_name: "test author" }],
      _links: {}
    });
  });
};

describe("Output/Listing", () => {
  it("renders properly", async () => {
    const { container, getByText } = render(
      <Store.Provider
        value={{
          publisher: publisher,
          api: api,
          tenants: [
            {
              name: 'tenant1', code: 'tenant1', routes: [
                { name: 'testroute1', id: 1 },
                { name: 'testroute2', id: 2 },
              ]
            },
            { name: 'tenant2', code: 'tenant2', routes: [] }
          ],

          actions: { setFilters: jest.fn() }
        }}
      >
        <Listing
          type="published"
          hide={false}
        />
      </Store.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
