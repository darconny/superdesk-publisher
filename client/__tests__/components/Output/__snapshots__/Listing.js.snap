// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output/Listing renders properly 1`] = `
<div
  class="sd-column-box__main-column relative"
>
  <div
    class="sd-loader"
  />
</div>
`;
